### DISCUSSION TITLE
Gracefully migrating a project with moor to sound null safety?

### DISCUSSION THREAD

**Discussion by mattsimgithub** [simolus3/drift#1489] [Q&A] [answered]
I'm trying to migrate a medium sized Flutter app to sound null safety, and the migrate tool is showing a large number of modifications in the moor_database.dart file and the moor_database.g.dart file. In moor_database.dart, I think every single column declaration has the column marked nullable, e.g.:

`  TextColumn? get breedName => text().withLength(min: 0, max: 150).nullable()() as Column<String>?;`

I'm not 100% sure this is wrong--maybe TextColumn is null for a while during object creation or something--but it seems wrong. 

If these columns can be left non-null, I'd like to do this systematically (rather than manually adding a whole bunch of migrate comment hints). _(although I would be willing to add a bunch of hints if that was the best answer, of course)_

I think there's some circularity here with the .g. file, which seems to be guiding the migrate tool on how the moor_database.dart file ought to be migrated. I could drop it entirely, but it's not advised to run the migrate tool while you have a bunch of dart analyze errors, of which there would be many.

So, I guess I have a couple questions, a specific one and a general one:

1) Am I right to think that column definition properties in Table classes should be non-null?
2) Do you have any recommendations for smoothing out the null migration process, specifically with regard to moor files?

**Comment by simolus3** [comment, marked as answer]
> I'm not 100% sure this is wrong--maybe TextColumn is null for a while during object creation or something--but it seems wrong.

It is wrong :) Column definitions refer to a column as a schematic entity, and that can't be null (even if the column holds a nullable value). The migration tool should give a reason for it deciding to make these nullable though, can you check that? Maybe I can change some things in drift/moor to help the migration tool.
Depending on how many other (valid) suggestions the tool gives for database files, you could also exclude those files from the tool and migrate them manually (if it's not too much effort). That may be easier than adding the hints everywhere.

> I could drop it entirely, but it's not advised to run the migrate tool while you have a bunch of dart analyze errors, of which there would be many.

I wouldn't recommend to delete the `.g.dart` file either. You can just let the migration tool do its thing, the next build run will override that file to what it's supposed to be.

One problem is that the migration tool may be a bit off about what to make nullable in your generated classes. The next build run would then introduce new nullability warnings in your app.

> Do you have any recommendations for smoothing out the null migration process, specifically with regard to moor files?

Not really I'm afraid. If your database files don't import lots of other code, you may want to consider a two-step migration:

1. Keep the `environment.sdk` at a version smaller than `2.12.0` in your pubspec, but add `//@dart=2.12` comments at the start of your database Dart code. It may not need many manual changes at all.
2. Re-run the build so that `moor_generator` generates null-safe `.g.dart` files
3. Now run the regular migration tool, which will give you better recommendations since the generated data and table classes have already been migrated.

**Comment by mattsimgithub** [comment]
> The migration tool should give a reason for it deciding to make these nullable though, can you check that? Maybe I can change some things in drift/moor to help the migration tool.

for line 86: `  TextColumn? get breedName => text().withLength(min: 0, max: 150).nullable()() as Column<String>?;`

the migration tool offers the following:

```
Edit Details
Changed type 'Column<String>' to be nullable at C:\work\flutter\service_tech_assistant\lib\database\moor_database.dart:86.

Nullability reason:
FlockHouses​.breedName (moor_database.dart:86): return type of FlockHouses​.breedName (moor_database​.dart:86:3)
Add /*?*/ hintAdd /*!*/ hint

FlockHouses​.breedName (moor_database.dart:86): assignment of dynamic value
FlockHouses​.breedName (moor_database.dart:86): dynamic type (moor_database​.dart:86:31)
FlockHouses​.breedName (moor_database.dart:86): dynamic type is nullable by definition

Actions:

Add /*!*/ hintAdd /*?*/ hint
```

Which doesn't mean much to me, but perhaps provides a hint to you.

> One problem is that the migration tool may be a bit off about what to make nullable in your generated classes. The next build run would then introduce new nullability warnings in your app.

Understood and thanks.

> Keep the environment.sdk at a version smaller than 2.12.0 in your pubspec, but add //@dart=2.12 comments at the start of your database Dart code. It may not need many manual changes at all.

I was not aware of this option, and I'm very intrigued by it. Thanks much. 

I'm about to step out of the office until Monday. Will respond then if you've got more questions of me. No more questions for you, and thanks for the reply.

**Comment by mattsimgithub** [comment]
I seem to have gotten to the other side of migrating moor_database.dart and moor_database.g.dart. Remarking on some weirdness in case it helps others:

I wanted to rely on the migration tool as much as possible, and to do this I ended up adding a lot of `/*!*/` hints to my table definitions to keep them non-nullable. I also accepted downcast hints like adding ` as Column<String>` to the end of certain columns, so I had a lot of columns that looked like this after accepting the revisions from the migration tool:

```
  TextColumn /*!*/ get ownerName =>
      text().withLength(min: 1, max: 150).nullable()() as Column<String>;
```

Warning, the moor generator parser **does not like these changes**.

When I tried generating the .g. file I got lots of parser errors, and, long story short, I ended up removing all of those non-nullable hints and all of the type casting that the migration tool added.

After removing those, the parser was happy, and it regenerated a nice, working moor_database.g.dart file keyed to Dart 2.12.

There were a number of other changes suggested by the migration tool that were useful (this is a regrettably huge file with lots of db methods mixed in with the table defs), so I don't mind having gone the long way around on the table defs. In fact I suspect it was necessary to do it this way for the migration tool to light up the other changes I kept. But it was a weird enough journey to be worth mentioning.
